Assets/Scripts/RedRunner/Characters/RedCharacter.cs

```diff
@@ -507,19 +529,42 @@ public override void Move ( float horizontalAxis )
 
 		public override void Jump ()
 		{
-			if ( !IsDead.Value && m_JumpsSoFar < m_AllowedJumps)
+			if (!IsDead.Value)
 			{
-				m_JumpsSoFar++;
-				Vector2 velocity = m_Rigidbody2D.velocity;
-				velocity.y = m_JumpStrength;
-				m_Rigidbody2D.velocity = velocity;
-				m_Animator.animator.ResetTrigger ( "Jump" );
-				m_Animator.SetTrigger ( "Jump" );
-				m_JumpParticleSystem.Play ();
-				AudioManager.Singleton.PlayJumpSound ( m_JumpAndGroundedAudioSource );
+				if (m_GroundCheck.IsGrounded || m_WallDetector.TouchingWall)
+				{
+					Vector2 velocity = m_Rigidbody2D.velocity;
+					velocity.y = m_JumpStrength;
+					m_Rigidbody2D.velocity = velocity;
+					m_Animator.animator.ResetTrigger("Jump");
+					m_Animator.SetTrigger("Jump");
+					m_JumpParticleSystem.Play();
+					AudioManager.Singleton.PlayJumpSound(m_JumpAndGroundedAudioSource);
+				}
+				else if (m_HasDoubleJump)
+				{
+					Vector2 velocity = m_Rigidbody2D.velocity;
+					velocity.y = m_DoubleJumpStrength;
+					m_Rigidbody2D.velocity = velocity;
+					m_Animator.animator.ResetTrigger("Jump");
+					m_Animator.SetTrigger("Jump");
+					m_JumpParticleSystem.Play();
+					AudioManager.Singleton.PlayJumpSound(m_JumpAndGroundedAudioSource);
+					m_HasDoubleJump = false;
```

Comment:
nit: refactor duplicated parts into helper (that takes a jump strength as param)

Reply:
true